From 428b31786ca37ab9f114d6c1c499eb5d73bdda80 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 13 Sep 2014 23:14:43 -0400
Subject: [PATCH] Configurable Keep Spawn Loaded range per world

This lets you disable it for some worlds and lower it for others.
It also bounds it to the view-distance, to not load more than needed like it currently does.
---
 src/main/java/net/minecraft/server/MinecraftServer.java | 6 +++---
 src/main/java/org/bukkit/craftbukkit/CraftServer.java   | 2 +-
 src/main/java/org/bukkit/craftbukkit/CraftWorld.java    | 5 +++--
 src/main/java/org/spigotmc/SpigotWorldConfig.java       | 4 ++++
 4 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 62d22ec..17a48e6 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -311,9 +311,9 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IMo
             ChunkCoordinates chunkcoordinates = worldserver.getSpawn();
             long j = ar();
             i = 0;
-
-            for (int k = -192; k <= 192 && this.isRunning(); k += 16) {
-                for (int l = -192; l <= 192 && this.isRunning(); l += 16) {
+            short radius = worldserver.spigotConfig.keepLoadedRange; // Spigot
+            for (int k = -radius; k <= radius && this.isRunning(); k += 16) { // Spigot
+                for (int l = -radius; l <= radius && this.isRunning(); l += 16) { // Spigot
                     long i1 = ar();
 
                     if (i1 - j > 1000L) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 6a27141..a018a23 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1018,7 +1018,7 @@ public final class CraftServer implements Server {
         System.out.print("Preparing start region for level " + (console.worlds.size() - 1) + " (Seed: " + internal.getSeed() + ")");
 
         if (internal.getWorld().getKeepSpawnInMemory()) {
-            short short1 = 196;
+            short short1 = internal.spigotConfig.keepLoadedRange; // Spigot
             long i = System.currentTimeMillis();
             for (int j = -short1; j <= short1; j += 16) {
                 for (int k = -short1; k <= short1; k += 16) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index b603521..02e4b13 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -1158,8 +1158,9 @@ public class CraftWorld implements World {
         int chunkCoordX = chunkcoordinates.x >> 4;
         int chunkCoordZ = chunkcoordinates.z >> 4;
         // Cycle through the 25x25 Chunks around it to load/unload the chunks.
-        for (int x = -12; x <= 12; x++) {
-            for (int z = -12; z <= 12; z++) {
+        int radius = world.spigotConfig.keepLoadedRange / 16; // Spigot
+        for (int x = -radius; x <= radius; x++) { // Spigot
+            for (int z = -radius; z <= radius; z++) { // Spigot
                 if (keepLoaded) {
                     loadChunk(chunkCoordX + x, chunkCoordZ + z);
                 } else {
diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index 854ba6d..badb3af 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -127,10 +127,14 @@ public class SpigotWorldConfig
     }
 
     public int viewDistance;
+    public short keepLoadedRange;
     private void viewDistance()
     {
         viewDistance = getInt( "view-distance", Bukkit.getViewDistance() );
         log( "View Distance: " + viewDistance );
+
+        keepLoadedRange = (short) (Math.min(getInt("keep-spawn-loaded-range", viewDistance), viewDistance) * 16);
+        log( "Keep Spawn Loaded Range: " + (keepLoadedRange/16));
     }
 
     public byte mobSpawnRange;
-- 
1.9.1

